src/lib/homestore_backend/hs_pg_manager.cpp

```diff
@@ -145,9 +145,55 @@ void HSHomeObject::on_create_pg_message_commit(int64_t lsn, sisl::blob const& he
     if (ctx) ctx->promise_.setValue(folly::Unit());
 }
 
-PGManager::NullAsyncResult HSHomeObject::_replace_member(pg_id_t id, peer_id_t const& old_member,
-                                                         PGMember const& new_member) {
-    return folly::makeSemiFuture< PGManager::NullResult >(folly::makeUnexpected(PGError::UNSUPPORTED_OP));
+PGManager::NullAsyncResult HSHomeObject::_replace_member(pg_id_t pg_id, peer_id_t const& old_member,
+                                                         PGMember const& new_member, uint32_t commit_quorum) {
+
+    group_id_t group_id;
+    {
+        auto lg = std::shared_lock(_pg_lock);
+        auto iter = _pg_map.find(pg_id);
+        if (iter == _pg_map.end()) return folly::makeUnexpected(PGError::UNKNOWN_PG);
+        auto& repl_dev = pg_repl_dev(*iter->second);
+
+        if (!repl_dev.is_leader() && commit_quorum == 0) {
+            // Only leader can replace a member
+            return folly::makeUnexpected(PGError::NOT_LEADER);
+        }
+        group_id = repl_dev.group_id();
+    }
+
+    return hs_repl_service()
+        .replace_member(group_id, old_member, new_member.id, commit_quorum)
+        .via(executor_)
+        .thenValue([this, pg_id, old_member, new_member](auto&& v) mutable -> PGManager::NullAsyncResult {
+            if (v.hasError()) { return folly::makeUnexpected(toPgError(v.error())); }
+
+            on_pg_replace_member(pg_id, old_member, new_member);
```

Comment:
The `on_pg_replace_member` here will only be execute on master which is wrong.

The member change is a nuraft config commit, we need to handle it in the `commit_config` callback(https://github.com/eBay/HomeStore/blob/0fa6a0c127680edb89291633ede8c937d10ba990/src/lib/replication/repl_dev/raft_state_machine.cpp#L204) and call up to HO, in HO we call the `on_pg_replace_member`.  That ensures the `on_pg_replace_member` be executed on every replica

Reply:
but if we do that , we will get callback in both adding new member as well as remove old member

Reply:
Ack